### ISSUE TITLE
ultralytics generate 3 times more thread than set in workers config

### ISSUE THREAD

**Issue by rootadminWalker** [ultralytics/ultralytics#18802]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I am training yolo11 small on my custom dataset with 10 workers, here is my train code:
```python
from ultralytics import YOLO

model = YOLO("weights/yolo11s.pt")

experiment = 'YOLOv11.small.imgsz832.Dataset'
project = f'./logs'

results = model.train(
    data='config.yaml',
    epochs=100,
    patience=10,
    imgsz=832,
    name=experiment,
    project=project,
    batch=8,
    workers=10
)
```

Ideally, this should generate 10 threads for data loading. However when I check with `ps x | grep python`, there are more than 33 threads running:
![Image](https://github.com/user-attachments/assets/3a49f128-3c18-4538-9e6e-567300e124d4)

When I debug the training program with gdb, it shows that between epochs, one thread is terminated and one thread is created
![Image](https://github.com/user-attachments/assets/d1697250-cfcb-4df4-9d0f-3876f10f3ba0)

Is it possible to explain why does this occurs? Thank you

**Comment by UltralyticsAssistant** [comment]
👋 Hello @rootadminWalker, thank you for your interest in Ultralytics 🚀! We appreciate you bringing up this issue. Let's see how we can help:

If this is a 🐛 Bug Report, we kindly ask you to provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us investigate the issue further. This includes:

1. The version of the `ultralytics` package you are using. Ensure you are on the latest version by running:
   ```bash
   pip install -U ultralytics
   ```
2. A simplified code snippet (or your current script) demonstrating the problem.
3. A description of your exact configuration, such as the Python version, operating system, and whether any virtual environments (conda, venv, etc.) are used.

---

### Upgrade

Please make sure you're using the latest `ultralytics` package alongside its [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml), installed in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify that your issue is not already resolved. You can upgrade using the following command:

```bash
pip install -U ultralytics
```

---

### Known Verified Environments

YOLO can be executed in various environments that have been pre-tested and verified, such as:

- **Notebooks** with free GPU:  
  <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM (`GCP`) - [Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI (`AWS`) - [Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image** - [Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/)  
  <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

---

### Resources

For further assistance, you can explore these resources:
- 📚 [Docs](https://docs.ultralytics.com/) for guides and examples
- ❓ [Model Training Tips](https://docs.ultralytics.com/guides/model-training-tips/) to optimize your training approach
- 🌎 [Community platforms](https://community.ultralytics.com/):  
  Join us on [Discord](https://discord.com/invite/ultralytics) 🎧, [Discourse](https://community.ultralytics.com/), or [Subreddit](https://reddit.com/r/Ultralytics)!

---

### Continuous Integration Status

Our CI system runs 24/7 on every commit and daily scheduled tests:  
<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>  

---

This is an automated response to help provide guidance. An Ultralytics engineer will be with you shortly to assist further! 😊

**Comment by Y-T-G** [comment]
> Ideally, this should generate 10 threads for data loading.

It would be 20. 10 for train dataloader and 10 for val dataloader.

What version of Ultralytics are you using?

**Comment by Y-T-G** [comment]
Actually it would be 30. 10 for train, and 20 for val.

https://github.com/ultralytics/ultralytics/blob/5306a8cc1f01ed79c8409b102df67b3cd90c717d/ultralytics/models/yolo/detect/train.py#L54

**Comment by rootadminWalker** [comment]
Hello, I'm using ultralytics 8.3.63
![Image](https://github.com/user-attachments/assets/8f56d0ab-90bf-41a7-a21b-68f725ecc93c)
To your explanation, l now understand that train and validation has their own dataloaders, but why train + val = 30? In the code your provide, double amount of workers are applied only if the mode is not 'train'. However, validation uses the validator, not the trainer, and the validator will use the exact amount of workers stated in the config. 

> Actually it would be 30. 10 for train, and 20 for val.
> 
> [ultralytics/ultralytics/models/yolo/detect/train.py](https://github.com/ultralytics/ultralytics/blob/5306a8cc1f01ed79c8409b102df67b3cd90c717d/ultralytics/models/yolo/detect/train.py#L54)
> 
> Line 54 in [5306a8c](/ultralytics/ultralytics/commit/5306a8cc1f01ed79c8409b102df67b3cd90c717d)
> 
>  workers = self.args.workers if mode == "train" else self.args.workers * 2

**Comment by Y-T-G** [comment]
`mode` would be `train` when it's creating the train dataloader and `val` when it's creating the val dataloader. Both are created during training. So 10 + 20.

**Comment by rootadminWalker** [comment]
Understand, this is very informative, thanks🙏
One last additional question, when debugging with gdb, one thread is continuously being killed and respawning at the end and the start of an epoch, do you know why this happens?

**Comment by Y-T-G** [comment]
Does it occur if val=False?
